src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncStateHandler.scala

```diff
@@ -0,0 +1,446 @@
+
+import io.iohk.ethereum.consensus.validators.Validators
+import io.iohk.ethereum.domain.{Account, BlockHeader, Blockchain, Receipt}
+import io.iohk.ethereum.network.Peer
+
+import scala.annotation.tailrec
+import FastSyncStateHandler._
+import akka.util.ByteString
+import io.iohk.ethereum.blockchain.sync.FastSync.SyncState
+import io.iohk.ethereum.blockchain.sync.FastSyncReceiptsValidator.ReceiptsValidationResult
+import io.iohk.ethereum.blockchain.sync.SyncBlocksValidator.BlockBodyValidationResult
+import io.iohk.ethereum.db.storage.AppStateStorage
+import io.iohk.ethereum.mpt.{BranchNode, ExtensionNode, LeafNode, MptNode}
+import io.iohk.ethereum.network.p2p.messages.PV62.BlockBody
+import io.iohk.ethereum.network.p2p.messages.PV63.NodeData
+import io.iohk.ethereum.utils.Config.SyncConfig
+import io.iohk.ethereum.utils.Logger
+import org.bouncycastle.util.encoders.Hex
+import io.iohk.ethereum.network.p2p.messages.PV63.MptNodeEncoders._
+
+import scala.util.{Failure, Success, Try}
+import io.iohk.ethereum.crypto.kec256
+
+
+class FastSyncStateHandler(
+  val blockchain: Blockchain,
+  val validators: Validators,
+  val syncConfig: SyncConfig,
+  val appStateStorage: AppStateStorage,
+  val randomFun: Int => Int
+) extends FastSyncReceiptsValidator with SyncBlocksValidator with Logger {
+
+  def handleHeaders(peer: Peer, headers: Seq[BlockHeader], requestedHeaders: BigInt)(syncState: SyncState): (FastSyncCommand, SyncState) = {
+    if (validHeaders(syncState ,headers, requestedHeaders))
+      handleBlockHeaders(syncState, peer, headers)
+    else
+      (ContinueSyncing(Some(BlackListCommand(peer, "malformed blockheaders"))), syncState)
+  }
+
+  def updateTargetBlock(state: FinalBlockProcessingResult)(syncState: SyncState): (FastSyncCommand, SyncState) = {
+    if (syncState.targetBlockUpdateFailures <= syncConfig.maximumTargetUpdateFailures) {
+      (ContTargetBlockUpdate(state), syncState.copy(updatingTargetBlock = true))
+    } else {
+      (AbortFastSync, syncState)
+    }
+  }
+
+  def handleNewTargetBlock(processState: FinalBlockProcessingResult,
+                           targetBlockHeader: BlockHeader)(syncState: SyncState): (FastSyncCommand, SyncState) = {
+    log.info(s"new target block with number ${targetBlockHeader.number} received")
+    if (targetBlockHeader.number >= syncState.targetBlock.number) {
+      val newState = updateTargetSyncState(syncState, processState, targetBlockHeader)
+      (BackToSyncing, newState.copy(updatingTargetBlock = false))
+    } else {
+      (AskForNewTarget(processState), syncState.copy(targetBlockUpdateFailures = syncState.targetBlockUpdateFailures + 1))
+    }
+  }
+
+
+  def handleBlockBodies(peer: Peer, requestedHashes: Seq[ByteString], blockBodies: Seq[BlockBody])(syncState: SyncState): (ContinueSyncing, SyncState) = {
+    if (blockBodies.isEmpty) {
+      val reason = s"got empty block bodies response for known hashes: ${requestedHashes.map(h => Hex.toHexString(h.toArray[Byte]))}"
```

Comment:
maybe you could make a helper method for extracting known hashes or for all reason message because your using `requestedHashes.map(h => Hex.toHexString(h.toArray[Byte]))` many times and the message itself differs mainly in what was empty

Reply:
tbh most of this methods was just extracted  without changes from `FastSync.scala`, and in this pr i would like to keep them as much similar as possible as this refactoring is already big. In the future if the time allows it would be good to extract similarities from `handleBlockBodies` , `handleReceipt`, `handleNodeData` as all of them follows the similar pattern.